redbot/core/utils/__init__.py

```diff
@@ -255,3 +257,128 @@ def bounded_gather(
     tasks = (_sem_wrapper(semaphore, task) for task in coros_or_futures)
 
     return asyncio.gather(*tasks, return_exceptions=return_exceptions)
+
+
+class AsyncIter(AsyncIterator[_T], Awaitable[List[_T]]):  # pylint: disable=duplicate-bases
+    """Asynchronous iterator yielding items from ``iterable`` that sleeps for ``delay`` seconds every ``steps`` items.
+
+    Parameters
+    ----------
+    iterable : Iterable
+        The iterable to make async.
+    delay: Union[float, int]
+        The amount of time in seconds to sleep.
+    steps: int
+        The number of iterations between sleeps.
+    """
+
+    def __init__(
+        self, iterable: Iterable[_T], delay: Union[float, int] = 0, steps: int = 1
+    ) -> None:
+        self._delay = delay
+        self._content = iter(iterable)
+        self._i = 0
+        self._steps = steps
+
+    def __aiter__(self) -> AsyncIter[_T]:
+        return self
+
+    async def __anext__(self) -> _T:
+        try:
+            item = next(self._content)
+        except StopIteration:
+            raise StopAsyncIteration
+        self._i += 1
+        if self._i % self._steps == 0:
+            await asyncio.sleep(self._delay)
+        return item
+
+    def __await__(self) -> Generator[Any, None, List[_T]]:
+        return self.flatten().__await__()
+
+    async def flatten(self) -> List[_T]:
+        """Returns a list of the iterable."""
+        return [item async for item in self]
+
+    def filter(self, function: Callable[[_T], Union[bool, Awaitable[bool]]]) -> AsyncFilter[_T]:
+        """
+        Filter the iterable with an (optionally async) predicate.
+
+        Parameters
+        ----------
+        function : Callable[[T], Union[bool, Awaitable[bool]]]
+            A function or coroutine function which takes one item of ``iterable``
+            as an argument, and returns ``True`` or ``False``.
+
+        Returns
+        -------
+        AsyncFilter[T]
+            An object which can either be awaited to yield a list of the filtered
+            items, or can also act as an async iterator to yield items one by one.
+
+        Examples
+        --------
+            >>> from redbot.core.utils import AsyncIter
+            >>> def predicate(value):
+            >>>    return value <= 5
```

Comment:
Let's use 4 space indent here
```suggestion
            >>>     return value <= 5
```